File: Application/c-c++/Set_Admetadata_To_Pipeline/set_classification_in_pipeline.cpp
// **
// Demo senario: 
// gst-launch-1.0 videotestsrc ! video/x-raw, format=BGR, width=640, height=480, framerate=30/1 ! adsetclassification ! admetadrawer ! videoconvert ! ximagesink

// This example only show how to get adlink metadata from appsink.
// So this example does not deal with any other detail concern about snchronize or other tasks.
// Only show how to retrieve the adlink metdata for user who is interested with them.
// **
#include "opencv2/opencv.hpp"
#include <iostream>
#include <stdio.h>
#include <thread>
#include <chrono>
#include <vector>
#include <gst/gst.h>

#include "gstadmeta.h"

using namespace cv;
using namespace std;

static GMainLoop *loop;
GstElement *pipeline, *videotestsrc, *filtercaps, *classifier, *drawer, *videoconvert, *appsink, *ximagesink;
vector<Mat> pipeLineOutputVec;
cv::Mat img(cv::Size(640, 480), CV_8UC3);

static void free_pipeline()
{
    gst_element_set_state (pipeline, GST_STATE_NULL);
    gst_object_unref (GST_OBJECT (pipeline));
    g_main_loop_unref (loop);
}
static void establish_thread_pipeline()
{
    /* init GStreamer */
    gst_init (NULL, NULL);
    loop = g_main_loop_new (NULL, FALSE);

    /* setup pipeline */
    pipeline = gst_pipeline_new ("pipeline");
    videotestsrc = gst_element_factory_make ("videotestsrc", "source");
    filtercaps = gst_element_factory_make ("capsfilter", "filtercaps");
    classifier = gst_element_factory_make ("adsetclassification", "classifier");
    drawer = gst_element_factory_make("admetadrawer", "drawer");
    videoconvert = gst_element_factory_make("videoconvert", "videoconvert");
    //appsink = gst_element_factory_make ("appsink", "appsink");
    ximagesink = gst_element_factory_make("ximagesink", "ximagesink");

    /* setup filtercaps*/
    g_object_set (G_OBJECT (filtercaps), "caps",
                  gst_caps_new_simple ("video/x-raw", "format", G_TYPE_STRING, "BGR",
                                       "width", G_TYPE_INT, 640, "height", G_TYPE_INT, 480,
                                       "framerate", GST_TYPE_FRACTION, 30, 1, NULL), NULL);
    gst_bin_add_many (GST_BIN (pipeline), videotestsrc, filtercaps, classifier, drawer, videoconvert, ximagesink, NULL);
    gst_element_link_many (videotestsrc, filtercaps, classifier, drawer, videoconvert, ximagesink, NULL);
    
    /* play */
    gst_element_set_state (pipeline, GST_STATE_PLAYING);
    g_main_loop_run (loop);
    
    free_pipeline();
}


int main(int, char**)
{
    thread pipethread(establish_thread_pipeline);
    cv::Mat img_show(cv::Size(640, 480), CV_8UC3);
    while(true)
    {
        if(pipeLineOutputVec.size() > 0)
        {
            img_show = pipeLineOutputVec[0].clone();
            imwrite("a.bmp", img_show);
            pipeLineOutputVec.erase(pipeLineOutputVec.begin());
        }
        this_thread::sleep_for(chrono::milliseconds(1));
    }
    
    return 0;
}

File: Application/python/Set_Admetadata_To_Pipeline/set_classification_in_pipeline.py
## **
## Demo senario: 
## gst-launch-1.0 videotestsrc ! video/x-raw, width=640, height=480 ! classifier_sample ! admetadrawer ! videoconvert ! ximagesink

## This example only show how to set adlink metadata.
## So this example does not deal with any other detail concern about snchronize or other tasks.
## Only show how to set the adlink metdata for user who is interested with them.
## **
import sys
import time
import numpy

# Required to import to set ADLINK inference metadata
import gst_admeta as admeta

import cv2
import gi
gi.require_version('Gst', '1.0')

from gi.repository import Gst, GObject

def on_message(bus: Gst.Bus, message: Gst.Message, loop: GObject.MainLoop):
    mtype = message.type
    
    if mtype == Gst.MessageType.EOS:
        print("End of stream")
        loop.quit()
    elif mtype == Gst.MessageType.ERROR:
        err, debug = message.parse_error()
        print("Gst.MessageType.ERROR catched in on_message:")
        print(err, debug)
        loop.quit()
    elif mtype == Gst.MessageType.ANY:
        err, debug = message.parse_warning()
        print(err, debug)

    return True

def establish_pipeline(pipeline, pipeline_elements):
    ## Add elements in pipeline.
    for element in pipeline_elements:
        pipeline.add(element)
    ## Link element one by one.
    for i in range(len(pipeline_elements) - 1):
        pipeline_elements[i].link(pipeline_elements[i + 1])


if __name__ == '__main__':
    print('Start establish pipeline.')
    # Initialize GStreamer
    Gst.init(sys.argv)

    # Create the elements
    ## element: videotesetsrc
    videosrc = Gst.ElementFactory.make("videotestsrc", "videosrc")
    
    ## element: capsfilter
    filtercaps = Gst.ElementFactory.make("capsfilter", "filtercaps")
    filtercaps.set_property("caps", Gst.Caps.from_string("video/x-raw, format=BGR, width=640, height=480"))

    ## element: classifier_sample
    classifier = Gst.ElementFactory.make("classifier_sample", "classifier")

    ## element: admetadrawer
    drawer = Gst.ElementFactory.make("admetadrawer", "drawer")

    ## element: videoconvert
    videoconvert = Gst.ElementFactory.make("videoconvert", "videoconvert")

    ## element: ximagesink
    sink = Gst.ElementFactory.make("ximagesink", "sink")
    
    # Create the empty pipeline
    pipeline = Gst.Pipeline().new("test-pipeline")
    
    # Build the pipeline
    pipeline_elements = [videosrc, filtercaps, classifier, drawer, videoconvert, sink]
    establish_pipeline(pipeline, pipeline_elements)

    # Start pipeline
    pipeline.set_state(Gst.State.PLAYING)
    loop = GObject.MainLoop()
    
    # Wait until error or EOS
    bus = pipeline.get_bus()
    bus.add_signal_watch()
    bus.connect("message", on_message, loop)

    try:
        print("Start to run the pipeline.\n")
        loop.run()
    except Exception:
        print("in exception")
        traceback.print_exc()
        loop.quit()

    # Stop Pipeline
    pipeline.set_state(Gst.State.NULL)
    del pipeline
    print('pipeline stopped.\n')
